### PR TITLE
Eagerness fixes around delete and create relationship

### PR DESCRIPTION
**Pull request by pontusmelke**
In several cases we were not eager enough, e.g.
- `MATCH (n)-[r]->(m) DELETE r,n,m`
- `MATCH (a)-[t:T]-(b) CREATE (a)-[:T]->(b) RETURN count(*) as count`
- `MATCH (n {prop : 5})-[r]-(m) SET m.prop = 5 RETURN count(*)`

Made sure tests all tests in `EagerizationAcceptanceTest` actually are executed,
and not just investigate the plans.

### CODE REVIEW

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/DeleteAcceptanceTest.scala [hunk 1]
```diff
@@ -71,4 +71,27 @@ class DeleteAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsT
 
     assertStats(result, nodesDeleted = 4, relationshipsDeleted = 3)
   }
+
+  test("expand, delete and count") {
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/DeleteAcceptanceTest.scala]: These tests are dependant on the direction of the match being `BOTH`, right? Perhaps we should include this in the test name.

File: community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/DeleteAcceptanceTest.scala [hunk 2]
```diff
@@ -71,4 +71,27 @@ class DeleteAcceptanceTest extends ExecutionEngineFunSuite with QueryStatisticsT
...
+    relate(createNode(), createNode())
+
+    val result = execute(s"MATCH (a)-[r]-(b) DELETE r,a,b RETURN count(*) AS c")
+    assertStats(result, nodesDeleted = 2, relationshipsDeleted = 1)
+
+    result.toList should equal(List(Map("c" -> 2)))
+  }
+
+  test("variable length expand, delete, and count") {
```
Comment (Mats-SX) [inline community/cypher/acceptance/src/test/scala/org/neo4j/internal/cypher/acceptance/DeleteAcceptanceTest.scala]: Same as above.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/addEagernessIfNecessary.scala [hunk 3]
```diff
@@ -65,29 +69,43 @@ object addEagernessIfNecessary extends (Pipe => Pipe) {
...
+  }
 
-  private def relsReadWriteConflict(from: Effects, to: Effects) = {
-    readsCreatesSameRelationship(from.regardlessOfOptionalEffects, to) || readsDeletesSameRelationship(from, to)
+  private def relsReadDeleteConflict(from: Effects, to: Effects) = {
+    readsDeletesSameRelationship(from, to)
   }
 
+  private def relsReadDeleteNodeConflict(from: Effects, to: Effects) =
+    from.regardlessOfOptionalEffects.containsRelationshipReads && to.contains(DeletesNode)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/addEagernessIfNecessary.scala:104]: Is this for `DETACH DELETE`? Perhaps mention in a comment.

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 4]
```diff
@@ -78,21 +114,27 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
         |MATCH (a)-[t:T]->(b)
         |DELETE t
         |MERGE (a)-[t2:T]->(b)
-        |RETURN t2.prop
+        |RETURN t2
       """.stripMargin
 
-    assertNumberOfEagerness(query, 1)
-    assertStats(updateWithBothPlanners(query), relationshipsDeleted = 2, relationshipsCreated = 1)
+    //TODO: We should not need eager between MATCH and DELETE?
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: This is true (no eager needed) for _directed_ rel matches, right?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 5]
```diff
@@ -302,10 +359,35 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
+  test("should not introduce an eager pipe between a non-directional leaf relationship read and a relationship create") {
+    relate(createNode(), createNode(), "TYPE")
+    relate(createNode(), createNode(), "TYPE")
+    val query = "MATCH (a)-[:TYPE]-(b) CREATE (a)-[:TYPE]->(b) RETURN count(*) as count"
+
+    //TODO this is probably safe, we need to check directions
+    val result = updateWithBothPlanners(query)
+    assertStats(result, relationshipsCreated = 4)
+    result.columnAs[Int]("count").next should equal(4)
+    assertNumberOfEagerness(query, 1)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:377]: Test name says `should not ... eager` but this asserts it should have... which is correct?

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 6]
```diff
@@ -302,10 +359,35 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
...
+
+  test("should not introduce an eager pipe between a non-directional leaf relationship read and a relationship merge") {
+    relate(createNode(), createNode(), "TYPE")
+    relate(createNode(), createNode(), "TYPE")
+    val query = "MATCH (a)-[:TYPE]-(b) MERGE (a)-[:TYPE]->(b) RETURN count(*) as count"
+
+    val result = executeWithRulePlanner(query)
+    assertStats(result, relationshipsCreated = 2)
+    //TODO: This should not need to be eager?
+    assertNumberOfEagerness(query, 1)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:388]: Test name and assertion misaligned

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 7]
```diff
@@ -126,7 +170,8 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
       """.stripMargin
 
     assertStats(executeWithRulePlanner(query), relationshipsDeleted = 2, relationshipsCreated = 1)
-    assertNumberOfEagerness(query, 1)
+    //TODO: We should not need eager between MATCH and DELETE?
+    assertNumberOfEagerness(query, 2)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: Test name and assertion misaligned

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 8]
```diff
@@ -109,7 +151,9 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
       """.stripMargin
 
     assertStats(executeWithRulePlanner(query), relationshipsDeleted = 2, relationshipsCreated = 1)
-    assertNumberOfEagerness(query, 0)
+    //TODO this might be safe without eager, given that the match is directed and probably need to check the direction
+    //TODO: We should not need eager between MATCH and DELETE?
+    assertNumberOfEagerness(query, 2)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala]: Test name and assertion misaligned

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala [hunk 9]
```diff
@@ -432,8 +580,7 @@ class EagerizationAcceptanceTest extends ExecutionEngineFunSuite with TableDrive
 
     val query = "MATCH (a:Person) OPTIONAL MATCH (a)-[r1]-() DELETE a, r1"
 
-    assertNumberOfEagerness(query, 0)
-    assertStats(updateWithBothPlanners(query), nodesDeleted = 2, relationshipsDeleted = 6)
```
Comment (Mats-SX) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:436]: Why was this line removed? I think we added it back when we cherry-picked into the other PR, @pontusmelke, but still it's strange.

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/addEagernessIfNecessary.scala (hunk 3 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/addEagernessIfNecessary.scala:104]: No it is for when we read relationships and delete nodes

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala (hunk 5 above)
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:377]: fixed

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala (hunk 6 above)
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:388]: fixed

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala (hunk 9 above)
Comment (pontusmelke) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/EagerizationAcceptanceTest.scala:436]: don't know what happened, not it is here and in the other PR at least!

File: community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/addEagernessIfNecessary.scala (hunk 3 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-3.0/src/main/scala/org/neo4j/cypher/internal/compiler/v3_0/executionplan/addEagernessIfNecessary.scala:104]: Oh. How does that require eagerness? I'm intrigued.
